pep-0669.rst

```diff
@@ -0,0 +1,318 @@
+
+Code objects can be instrumented by calling::
+
+  instrumentation.instrument(codeobject, events)
+
+Code objects must be instrumented before they are executed.
+An exception will be raised if the code object has been executed before it
+is instrumented.
+
+Register callback functions for instrumentation
+'''''''''''''''''''''''''''''''''''''''''''''''
+
+To register a callable for events call::
+
+  instrumentation.register(event, func)
+
+Functions can be unregistered by calling
+``instrumentation.register(event, None)``.
+
+Callback functions can be registered at any time.
+
+Callback function arguments
+'''''''''''''''''''''''''''
+
+When an event occurs the registered function will be called.
+The arguments provided are as follows:
+
+* BRANCH: ``func(code: CodeType, offset: int, taken:bool)``
+* JUMPBACK: ``func(code: CodeType, offset: int)``
+* ENTER: ``func(code: CodeType, offset: int)``
+* EXIT: ``func(code: CodeType, offset: int)``
+* C_CALL: ``func(code: CodeType, offset: int, value: object)``
+* C_RETURN: ``func(code: CodeType, offset: int, value: object)``
+* C_EXCEPT: ``func(code: CodeType, offset: int, exception: BaseException)``
+* RAISE: ``func(code: CodeType, offset: int, exception: BaseException)``
+* EXCEPT: ``func(code: CodeType, offset: int)``
+* UNWIND: ``func(code: CodeType)``
+
+Monitoring
+----------
+
+Monitoring allows checkpoints to be inserted or removed at any
+point in the program's execution.
+
+The following functions are provided to insert monitoring points::
+
+  instrumentation.insert_monitors(codeobject, *offsets)
+  instrumentation.remove_monitors(codeobject, *offsets)
+  instrumentation.monitor_off(codeobject, offset)
+  instrumentation.monitor_on(codeobject, offset)
+
+All functions return ``True`` if a monitor checkpoint was present,
+or ``False`` if a monitor checkpoint was not present.
+Turning on, or off, a non-existent checkpoint is a no-op;
+no exception is raised.
+
+To register a callable for monitoring function events call::
+
+  instrumentation.monitor_register(func)
+
+The callback function will be called with the code object and offset as arguments::
+
+  func(code: CodeType, offset: int)
+
+For optimizing virtual machines, such as future versions of CPython
+(and ``PyPy`` should they choose to support this API), a call to
+``insert_monitors`` and ``remove_monitors`` in a long running program
+could be quite expensive, possibly taking 100s of milliseconds as it
+triggers de-optimizations. Repeated calls to ``insert_monitors``
+and ``remove_monitors``, as may be required in an interactive debugger,
+should be relatively inexpensive.
+
+Combining Checkpoints
+---------------------
+
+Only one instrumentation checkpoint and one monitoring checkpoint is allowed
+per bytecode instruction. It is possible to have both a monitoring and
+instrumentation checkpoint on the same instruction; they are independent.
+Monitors will be called before instrumentation if both are present.
```

Comment:
This confused me, since the instrumentation API has no explicit mention of checkpoints (there are no offsets in the instrumentation API, only code objects). I take it that instrumentation works by replacing specific bytecodes (e.g. certain JUMP instructions for BRANCH and JUMPBACK), and monitoring *also* works by doing that? And the instruction has two flag bits indicating whether it is an instrumentation or monitoring checkpoint or both?

It would seem that the APIs already imply that only one instrumentation checkpoint can exist per instruction, and only one monitoring checkpoint. So this paragraph only adds in which order they will be called.